src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp

```diff
@@ -54,42 +37,119 @@ APPLICATION_INFO::~APPLICATION_INFO()
 HRESULT
 APPLICATION_INFO::Initialize(
     _In_ IHttpServer              *pServer,
-    _In_ IHttpApplication         *pApplication,
-    _In_ FILE_WATCHER             *pFileWatcher
+    _In_ IHttpApplication         *pApplication
 )
 {
     HRESULT hr = S_OK;
 
     DBG_ASSERT(pServer);
     DBG_ASSERT(pApplication);
-    DBG_ASSERT(pFileWatcher);
 
     // todo: make sure Initialize should be called only once
     m_pServer = pServer;
     FINISHED_IF_NULL_ALLOC(m_pConfiguration = new ASPNETCORE_SHIM_CONFIG());
     FINISHED_IF_FAILED(m_pConfiguration->Populate(m_pServer, pApplication));
     FINISHED_IF_FAILED(m_struInfoKey.Copy(pApplication->GetApplicationId()));
-    FINISHED_IF_NULL_ALLOC(m_pFileWatcherEntry = new FILE_WATCHER_ENTRY(pFileWatcher));
 
-    UpdateAppOfflineFileHandle();
+    //UpdateAppOfflineFileHandle();
 
 Finished:
     return hr;
 }
 
-HRESULT
-APPLICATION_INFO::StartMonitoringAppOffline()
+BOOL
+APPLICATION_INFO::AppOfflineFound()
 {
-    if (m_pFileWatcherEntry != NULL)
+    ULONGLONG  ulCurrentTime = GetTickCount64();
+    //
+    // we only care about app offline presented. If not, it means the application has started
+    // and is monitoring  the app offline file
+    // we cache the file exist check for 1 minute
+    //
+    if (m_fAppOfflineFound && ulCurrentTime - m_ulLastCheckTime > 1000)
     {
-        RETURN_IF_FAILED(m_pFileWatcherEntry->Create(m_pConfiguration->QueryApplicationPhysicalPath()->QueryStr(), L"app_offline.htm", this, NULL));
+        // double check to avoid expensive IO operation
+        SRWExclusiveLock lock(m_srwLock);
+        if (ulCurrentTime - m_ulLastCheckTime > 1000)
+        {
+            UpdateAppOfflineFileHandle();
+            m_ulLastCheckTime = ulCurrentTime;
+        }
     }
-    return S_OK;
+    return m_fAppOfflineFound;
+}
+
+// Load appoffline content
+BOOL APPLICATION_INFO::LoadAppOffline(LPWSTR strFilePath)
+{
+    BOOL            fResult = TRUE;
+    LARGE_INTEGER   li = { 0 };
+    CHAR           *pszBuff = NULL;
+    HANDLE         handle = INVALID_HANDLE_VALUE;
+
+    DBG_ASSERT(strFilePath);
+
+    handle = CreateFile(strFilePath,
```

Comment:
Handle wrapped and logging in this method.

Reply:
I did it in https://github.com/aspnet/IISIntegration/pull/977/files#diff-0ecd7731bba35e311fac903916b30e44R26